tiered-storage/file-system/src/main/java/org/apache/bookkeeper/mledger/offload/filesystem/impl/FileSystemManagedLedgerOffloader.java

```diff
@@ -0,0 +1,241 @@
+                        MapFile.Writer.keyClass(LongWritable.class),
+                        MapFile.Writer.valueClass(BytesWritable.class));
+                key.set(-1);
+                byte[] ledgerMetadata = buildLedgerMetadataFormat(readHandle.getLedgerMetadata());
+                value.set(buildLedgerMetadataFormat(readHandle.getLedgerMetadata()), 0, ledgerMetadata.length);
+                dataWriter.append(key, value);
+                long haveOffloadEntryNumber = 0;
+                do {
+                    long end = Math.min(haveOffloadEntryNumber + ENTRIES_PER_READ - 1, readHandle.getLastAddConfirmed());
+                    LedgerEntries ledgerEntriesOnce = readHandle.readAsync(haveOffloadEntryNumber, end).get();
+                    log.debug("read ledger entries. start: {}, end: {}", haveOffloadEntryNumber, end);
+                    Iterator<LedgerEntry> iterator = ledgerEntriesOnce.iterator();
+                    while (iterator.hasNext()) {
+                        LedgerEntry entry = iterator.next();
+                        long entryId = entry.getEntryId();
+                        key.set(entryId);
+                        value.set(entry.getEntryBytes(), 0, entry.getEntryBytes().length);
+                        dataWriter.append(key, value);
+                        haveOffloadEntryNumber ++;
+                    }
+                } while (haveOffloadEntryNumber - 1 < readHandle.getLastAddConfirmed());
+                if (haveOffloadEntryNumber - 1 != readHandle.getLastAddConfirmed()) {
+                    log.error("The expected number of entries in offload does not match the actual number");
+                    throw new IOException();
+                }
+                IOUtils.closeStream(dataWriter);
+                promise.complete(null);
+            } catch (InterruptedException | ExecutionException | IOException e) {
+                log.error("Exception when get CompletableFuture<LedgerEntries>. ", e);
+                if (e instanceof InterruptedException) {
+                    Thread.currentThread().interrupt();
+                }
+                promise.completeExceptionally(e);
+                return;
+            }
+        });
+        return promise;
+    }
+
+    @Override
+    public CompletableFuture<ReadHandle> readOffloaded(long ledgerId, UUID uuid, Map<String, String> offloadDriverMetadata) {
+
+        CompletableFuture<ReadHandle> promise = new CompletableFuture<>();
+        String storagePath = getStoragePath(storageBasePath, offloadDriverMetadata.get("ManagedLedgerName"));
+        String dataFilePath = getDataFilePath(storagePath, ledgerId, uuid);
+        scheduler.chooseThread(ledgerId).submit(() -> {
+            try {
+                MapFile.Reader reader = new MapFile.Reader(new Path(dataFilePath),
+                        configuration);
+                promise.complete(FileStoreBackedReadHandleImpl.open(scheduler.chooseThread(ledgerId), reader, ledgerId));
```

Comment:
It seems that we are sharing a scheduler for both reading and writing. If a ledger is offloading, then it will be blocked the reading operations. Can we use different executors for reading and writing?

Reply:
I'll think it over :)